### PR TITLE
Happy Eyeballs: Do not discard viable reforwarding destinations

### PR DESCRIPTION
**Pull request by eduard-bagdasaryan**
When HappyConnOpener starts opening two connections, both destinations
are removed from the shared destinations list. As soon as one connection
(X) succeeded, the other destination (Y) was essentially forgotten.  If
FwdState, after using X, decided to reforward the request, then the
request was never reforwarded to Y. We now return Y to the list.

Also abort the still-active ConnOpener job upon the termination of its
"parent" HappyConnOpener job. Without an explicit termination, those
abandoned child jobs wasted OS and Squid resources while the OS was
trying to open the requested connection. They would terminate on a
timeout or when the connection was finally opened (and discarded).

Waiting for the child ConnOpener job to end is a useful optimization,
but it remains a TODO. It requires complex accumulation avoidance logic.

### CODE REVIEW

**Comment by eduard-bagdasaryan** [comment]
> It looks suspicious that we call cancelAttempt() only in one specific job outcome.

This method should probably have been named 'canchelAttemptAndRetry()' or smth. like that. The point is that we retry the path only on that specific condition - when one connection succeeds (as the PR description outlines). In other cases, we should only cancel pending connections without attempting to retry them (for otherwise we would end up in endless re-attempting stalled connections). We could remove cancelAttempt() call from HappyConnOpener::connectDone() and invoke it from swanSong() (for both connections), but only after checking  an additional condition - i.e., that one of the connections succeeded (and another is pending). I do not see how this can be checked in swanSong() (i.e., that params.flag == Comm::OK) without introducing a new variable). If so, we should leave cancelAttempt() call as is, and probably rename the method.

**Comment by rousskov** [comment]
> we retry the path only on that specific condition

Why not return the path in _all_ cases where we have not finished trying that path?

> only on that specific condition when one connection succeeds

AFAICT, that is not exactly true (or is misleading) because sendSuccess() is also called in another case where we do not call cancelAttempt().

**Comment by eduard-bagdasaryan** [comment]
> Why not return the path in all cases where we have not finished trying that path?

We can, if we are sure that this retrying will not last forever (i.e., a stalled connection was canceled and retried again and again). This should be true if at least one of the two started ConnOpener job returns earlier than the parent HappyConnOpener job.

**Comment by eduard-bagdasaryan** [comment]
> AFAICT, that is not exactly true (or is misleading) because sendSuccess() is also called in another case

This another case is in reuseOldConnection(). My understanding is that if there is a suitable old connection which we reuse and immediately return, we do not even try the spare connection and, therefore, we have nothing to cancel/retry. Am I wrong?

**Comment by rousskov** [comment]
> My understanding is that if there is a suitable old connection which we reuse and immediately return, we do not even try the spare connection and, therefore, we have nothing to cancel/retry.

Please correct me if I am wrong, but my understanding is that we do not check whether the _spare_ connection is available from the persistent connection pool until we are allowed to open a spare connection (which happens some time after we start opening a prime connection and while we are still trying to open a prime connection).

That lack of "look ahead" optimization for spare persistent connections is unfortunate, but correctly implementing that optimization (e.g., without violating path use order and without needlessly killing off persistent connections) may not be trivial and, hence, is outside of this issue scope unless it will significantly simplify this issue code as well.

**Comment by rousskov** [comment]
> we do not even try the spare connection and, therefore, we have nothing to cancel/retry.

To clarify the implication of what I said in previous comment: We have to cancel/retry the _prime_ connection in this case.

**Comment by rousskov** [comment]
> > Why not return the path in all cases where we have not finished trying that path?
 
> We can, if we are sure that this retrying will not last forever (i.e., a stalled connection was canceled and retried again and again).

If we cancel only when the HappyConnOpener job ends, then there can be no "forever" as far as HappyConnOpener code is concerned. If FwdState or tunnel.cc retry forever, then that is their bug. Checking for and fixing that bug, if it exists, is in this issue scope. I hope that forward_max_tries (and other considerations discussed below) will prevent any forever retries.

> This should be true if at least one of the two started ConnOpener job returns earlier than the parent HappyConnOpener job.

If we cancel only when the HappyConnOpener job ends, then either the job found/returned a usable connection or there was an error/problem. In the former case, we are making progress. In the latter case, no progress is possible, and FwdState or tunnel.cc will (or has) quit. In both cases, there is no "forever" looping.

Multiple retries of the same path will become possible, but AFAICT, we _welcome_ at least the first such retry because without it we have not fixed the bug this PR is trying to fix. It could be argued that repeated retries are not welcomed in some cases, but I do not see how we can naturally prevent them without adding more configuration options. In other words, I do not see why the second retry should be prohibited just because we have tried to open a connection to that destination twice already and had to cancel every time. Yes, that troubled path is apparently slower than other alternatives, but those fast alternatives are clearly not working well either or we would not be reforwarding after finding them.

**Comment by eduard-bagdasaryan** [comment]
> In the former case, we are making progress.

Agreed.

> In the latter case, no progress is possible,

But what shall we do here (i.e., in HappyConnOpener::swanSong()) if both ConnOpener jobs are still executing? Or such a situation is impossible (i.e., at least one of them should at least return an error by that time)? If we retry both connections, on the next iteration they will be tried again and so on, so on.

**Comment by eduard-bagdasaryan** [comment]
> we do not check whether the spare connection is available from the persistent connection pool until we are allowed to open a spare connection (which happens some time after we start opening a prime connection and while we are still trying to open a prime connection).

I think you are right: I misread the code (assuming that it was optimized to check for available cached connections first before starting to open a new one).

> In the former case, we are making progress.

Agreed.

>  In the latter case, no progress is possible, and FwdState or tunnel.cc will (or has) quit. 

I agree that there is no 'lopping' (FwdState sets flags.dont_retry=true and exits without further retrying). If so, it is useless to re-add these destinations (i.e., call our cancelAttempt() as in the first case) because the parent FwdState will not use them.

So we have two alternatives:
1. Add the missing cancelAttempt() before another sendSuccess() (or rework and move this method into sendSuccess()).
2. Call cancelAttempt() from HappyConnOpener::swanSong() for both connections (i.e., instead of each Attempt::cancel()). If one of the connections was successful, the result will be as (1). If both connections were unsuccessful, we uselessly re-add connections into destinations, but this will make no harm since FwdState will exit soon without retrying.

**Comment by rousskov** [comment]
> So we have two alternatives:
1. Try to remember to call cancelAttempt() when it may be needed.
2. Always call cancelAttempt() (in case it might be needed).

I think the second alternative is the only viable one. I would rephrase it as "we must cancel any unfinished attempts before forgetting about them". The swanSong() method already has related `if` statements for each of the two possibly pending attempts. We need to refactor the bodies of those `if` statements to call cancelAttempt() AFAICT. 

It would be good to make it impossible to simply forget the previous ConnOpener job except during the HappyConnOpener job destruction (swanSong() will protect the destruction path). I have not studied how hard it would be to eliminate such mistakes by adjusting Attempt API.

Please make sure you cancel the spare attempt first to avoid reordering paths.

Yes, in some cases, we will return an address that will never be used, but it is the right thing to do from code logic point of view, and the performance overhead of retryPath() is not very large. We should consider replacing std::vector in ResolvedPeers with std::deque though because std::deque has O(1) front insertions. Do common std::deque implementations make its construction approximately as expensive as std::vector construction? Or is there significant penalty for constructing a small std::deque compared to a small std::vector.

One could try to argue that, in the common case, ResolvedPeers probably have one or two addresses (IPv6 and IPv4), and we will be returning only one address (e.g., IPv6), making retryPath() with std::vector essentially an O(1) operation because there is nothing in that vector that needs shifting at swanSong() time. However, I am not sure the "one or two addresses" premise is valid. One could also argue that in common cases of popular domains, ResolvedPeers will have a few addresses (e.g., www.google.com had 7 IPv4 and IPv6 addresses in my test). With 5-10 addresses, will std::deque construction and one retryPath() call be cheaper than std::vector construction and one retryPath() call?

**Comment by eduard-bagdasaryan** [comment]
> Do common std::deque implementations make its construction approximately as expensive as std::vector construction?

std::deque [documentation](https://en.cppreference.com/w/cpp/container/deque) states that there is an overhead on accessing elements and also "large minimal memory cost". In my test, constructing a deque with 1 element (long int) led to 2 memory allocations (chunks of 64 and 512 bytes accordingly), whereas vector allocated just 8 bytes, as expected. If we append elements, deque grows by 512 chunks, while vector increases its capacity as 2**n (2,4,8, etc). Erasing an element from the vector's front, as expected, caused (n-1) copies (via operator=). So the benefit of using deque instead of vector for storing several(~10) small elements is vague: we optimize few memory copies but consume far more memory than needed in most cases and do two initial allocations instead of one.

**Comment by rousskov** [comment]
> Alex: We should consider replacing std::vector in ResolvedPeers with std::deque though because std::deque has O(1) front insertions

> Eduard: constructing a deque with 1 element (long int) led to 2 memory allocations

Thank you for checking this. Let's not switch to using std::deque -- guaranteed additional memory allocations are not worth the occasional push_front() speed savings. If we decide to optimize this later, we can simply add a "next element" iterator to the existing (private) std::vector-based storage.

**Comment by eduard-bagdasaryan** [comment]
> I think the second alternative is the only viable one

Done (c64784f).

File: src/HappyConnOpener.cc [hunk 1]
```diff
@@ -397,15 +397,11 @@ HappyConnOpener::swanSong()
     // TODO: Find an automated, faster way to kill no-longer-needed jobs.
 
     if (prime) {
-        if (prime.connector)
-            prime.connector->cancel("HappyConnOpener object destructed");
-        prime.clear();
+        cancelAttempt(prime, "HappyConnOpener object destructed");
```
Comment (rousskov) [inline src/HappyConnOpener.cc]: Rule of thumb: Avoid identical description of different contexts. Unique descriptions simplify triage.

```suggestion
        cancelAttempt(prime, "job finished during a prime attempt");
```

File: src/HappyConnOpener.cc [hunk 2]
```diff
@@ -397,15 +397,11 @@ HappyConnOpener::swanSong()
...
     }
 
     if (spare) {
-        if (spare.connector)
-            spare.connector->cancel("HappyConnOpener object destructed");
-        spare.clear();
+        cancelAttempt(spare, "HappyConnOpener object destructed");
```
Comment (rousskov) [inline src/HappyConnOpener.cc]:
```suggestion
        cancelAttempt(spare, "job finished during a spare attempt");
```

File: src/HappyConnOpener.cc [hunk 3]
```diff
@@ -869,3 +876,14 @@ HappyConnOpener::ranOutOfTimeOrAttempts() const
...
 }
 
+void
+HappyConnOpener::Attempt::cancel(const char *reason)
+{
+    if (connector)
+        connector->cancel(reason);
+    if (connOpener.valid()) {
+        CallJobHere(17, 3, connOpener, Comm::ConnOpener, noteAbort);
+    }
```
Comment (rousskov) [inline src/HappyConnOpener.cc]: If having a non-nil `connector` implies that connOpener is set (but not necessarily valid), then I would simplify the above:

```suggestion
    if (connector) {
        connector->cancel(reason);
        CallJobHere(17, 3, connOpener, Comm::ConnOpener, noteAbort);
    }
```

IIRC, CallJobHere() will take care of pointers to gone jobs (i.e., it will work correctly when connOpener.valid() is false).

**Comment by rousskov** [comment]
Closing in favor of the official PR 567.
